coq-core.opam

```diff
@@ -0,0 +1,34 @@
+synopsis: "The Coq Proof Assistant"
+description: """
+Coq is a formal proof management system. It provides
+a formal language to write mathematical definitions, executable
+algorithms and theorems together with an environment for
+semi-interactive development of machine-checked proofs. Typical
+applications include the certification of properties of programming
+languages (e.g. the CompCert compiler certification project, or the
+Bedrock verified low-level programming library), the formalization of
+mathematics (e.g. the full formalization of the Feit-Thompson theorem
+or homotopy type theory) and teaching.
+"""
+opam-version: "2.0"
+maintainer: "The Coq development team <coqdev@inria.fr>"
+authors: "The Coq development team, INRIA, CNRS, and contributors."
+homepage: "https://coq.inria.fr/"
+bug-reports: "https://github.com/coq/coq/issues"
+dev-repo: "git+https://github.com/coq/coq.git"
+license: "LGPL-2.1"
+
+version: "dev"
+
+depends: [
+  "ocaml"     { >= "4.05.0" }
+  "dune"      { >= "2.5.0"  }
```

Comment:
In the Coq Call, it was said that this lower bound could be made even lower (for the pure OCaml coq-core package). But now that I think of it, this would require a separate `dune-project` file for `coq-core` and for `coq-stdlib`, since this is the file that specifies the minimal Dune version to use. That's a shame because I'm not even sure that this is a supported use case.

Reply:
Due to #13437, I would on the contrary say that the bound should be put even higher, i.e., 2.8. Because it certainly does not work for me (unless something changed since I reported that bug).

Reply:
This bug should be closed now.

Note that the issue with 2.8 is that it is very recent and not yet packaged in Debian / Ubuntu, which would break @JasonGross's ppa, which is currently important for several projects' CI.

The plan laid out by @ejgallego during the call yesterday was that Dune would be required only to build the OCaml parts of Coq and there would be a choice of a coq_makefile-based or a Dune-based build for the standard library. If in addition, we could require different Dune lower bounds for the two parts (but I'm worried of this not being technically feasible), this plan would present the nice additional property of not forcing any recent Dune version upon our users.

Reply:
Yes we can uncouple the versions but that would require using different `dune-project` files for coq-core and coq-stdlib, so that'd be a further refinement.

Reply:
I think this worth it, but indeed, it can be done later. @JasonGross: you have no problem with Dune 2.5 becoming a requirement to build Coq, right?

Reply:
Ah, hold on, I should have looked into this more carefully earlier.  I've managed to package Dune 2.7 with OCaml 4.08.1 on all of the versions of Ubuntu that I care about, but I'm concerned that the 10x slowdown @ppedrot found at https://github.com/ocaml/ocaml/issues/7826#issuecomment-625342743 will hit fiat-crypto.  So it looks like I may need to move to 4.11.1 (can someone confirm that this version of OCaml does not suffer from this issue?), and I haven't yet packaged 4.11.1 for the various distributions I care about.  (And I may run into snags, as I did in packaging 4.08, such as OCaml being incompatible with the distributed version of binutils.)

Reply:
What about using Dune 2.7 compiled with OCaml 4.08.1 but staying on OCaml 4.07.1 for compiling Coq? Is this doable in your environment? If yes, that would be the safest solution.

Reply:
Indeed, OCaml 4.07 was release almost 3 years ago so it is safe start preparing the move from that version, as the GC parameter bug that was creating some problem has been finally solved, it is likely the recommended version becomes 4.12.0 soon.

We already discussed it, but it seems to me that using Debian packages to track Coq's master is gonna always be quite a bit of work, and will get worse in the future as document managers and other tooling are allowed to depend in even more packages. IMHO other solutions such as Nix / Docker / Opam are much preferable and tend to work out of the box. It makes little sense for the Coq team to have to worry about the complexities of Debian packaging as it is really a distribution issue, not related at all to Coq development itself so hard to justify to spend significant cycles on.

Reply:
[OCaml 4.11.1 does not even build on Groovy (Ubuntu 20.10), because many tests fail](https://launchpadlibrarian.net/525230922/buildlog_ubuntu-groovy-i386.ocaml_4.11.1-4.1~groovy~ppa1_BUILDING.txt.gz) with things like:
```
Running tests from 'tests/regression/pr3612' ...
 ... testing 'pr3612.ml' with 1 (native) => failed (The file /<<PKGBUILDDIR>>/testsuite/_ocamltest/tests/regression/pr3612/pr3612/ocamlopt.byte/ocamlopt.byte.output was expected to be empty because there is no reference file /<<PKGBUILDDIR>>/testsuite/tests/regression/pr3612/pr3612.compilers.reference but it is not:
========================================
/usr/bin/ld: pr3612.o: warning: relocation in read-only section `.text'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
========================================
```
What packages need to be upgraded or what needs to be done to fix this issue?

Reply:
No idea @JasonGross , I'm sorry, I wish I could be of more help but unfortunately I'm totally booked these days.

But that's a bit my point, very unfortunately, we don't have the resources in the Coq team these to be able to promise to help with there Coq-unrelated issues on a timely manner :( And on the other hand, our tooling / build systems needs a much needed refresh which we cannot delay more and will actually solve several dozen of issues with the current setup [such as rebases / bisects building properly, quick compilation, compilation caching, reduced maintenance, modularity, etc...]

I'd suggest tho you get in touch with Ralf Treinen and discuss.ocaml if you need Debian + OCaml help as I'm sure they will be much more helpful that we can [there is also OCaml discord and slack which is quite active]

Reply:
The build failure is surprising tho as the Debian packages for 4.11.1 work well; likely you need to update binutils, but this is a good example of why this crosses Coq's boundaries, updating binutils does impact almost _all_ packages in a Debian system so that is way out of our league.

Personally I'd abandon Debian packages as a CI platform and switch to Docker / Nix.

Reply:
Strangely enough, the build works fine in trusty, xenial, bionic, and focal (all of which are older than groovy) and hisrute (which is newer than groovy). Since groovy EOL's in July 2021, I can probably ignore this failure.

Reply:
I've made a post on [ocaml.discuss](https://discuss.ocaml.org/t/building-ocaml-4-11-1-on-ubuntu-20-10/7370/1), though it got hidden by a spam filter.  Is this the best platform for this?

Reply:
> I've made a post on [ocaml.discuss](https://discuss.ocaml.org/t/building-ocaml-4-11-1-on-ubuntu-20-10/7370/1), though it got hidden by a spam filter. Is this the best platform for this?

Ups ! I guess a moderator will take of it soon and approve your account, but indeed it is the best place to ask IMHO.